# Third: Evaluate Data Quality Issues in the Data Provided
Using the programming language of your choice (SQL, Python, R, Bash, etc...) identify as many data quality issues as you can. We are not expecting a full blown review of all the data provided, but instead want to know how you explore and evaluate data of questionable provenance.

Commit your code and findings to the git repository along with the rest of the exercise.

## Receipts

In [3]:
# Import necessary libraries and upload receipts data
import pandas as pd
from google.colab import files

uploaded = files.upload()
filename = list(uploaded.keys())[0]  # Geting the file name


receipts_df = pd.read_csv(filename)
receipts_df.head()

Saving receipts.csv to receipts (1).csv


receipt_id                   user_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575  5ff1e1eacfcf6c399c274ae6              500.0   
1  5ff1e1bb0a720f052300056b  5ff1e194b6a9d73a3a9f1052              150.0   
2  5ff1e1f10a720f052300057a  5ff1e1f1cfcf6c399c274b0b                5.0   
3  5ff1e1ee0a7214ada100056f  5ff1e1eacfcf6c399c274ae6                5.0   
4  5ff1e1d20a7214ada1000561  5ff1e194b6a9d73a3a9f1052                5.0   

                             bonusPointsEarnedReason  createDate_epoch  \
0  Receipt number 2 completed, bonus point schedu...     1609687531000   
1  Receipt number 5 completed, bonus point schedu...     1609687483000   
2                         All-receipts receipt bonus     1609687537000   
3                         All-receipts receipt bonus     1609687534000   
4                         All-receipts receipt bonus     1609687506000   

                createDate  dateScanned_epoch              dateScanned  \
0  2021-01-03 15:25:31.000      1609687531000  2021-01-03 15:25:31.000   
1  2021-01-03 15:24:43.000      1609687483000  2021-01-03 15:24:43.000   
2  2021-01-03 15:25:37.000      1609687537000  2021-01-03 15:25:37.000   
3  2021-01-03 15:25:34.000      1609687534000  2021-01-03 15:25:34.000   
4  2021-01-03 15:25:06.000      1609687506000  2021-01-03 15:25:06.000   

   finishedDate_epoch             finishedDate  ModifyDate_epoch  \
0        1.609688e+12  2021-01-03 15:25:31.000     1609687536000   
1        1.609687e+12  2021-01-03 15:24:43.000     1609687488000   
2                 NaN                      NaN     1609687542000   
3        1.609688e+12  2021-01-03 15:25:34.000     1609687539000   
4        1.609688e+12  2021-01-03 15:25:11.000     1609687511000   

                modifyDate  pointsAwardedDate_epoch    pointsAwardedDate  \
0  2021-01-03 15:25:36.000             1.609688e+12  2021-01-03 15:25:31   
1  2021-01-03 15:24:48.000             1.609687e+12  2021-01-03 15:24:43   
2  2021-01-03 15:25:42.000                      NaN                  NaN   
3  2021-01-03 15:25:39.000             1.609688e+12  2021-01-03 15:25:34   
4  2021-01-03 15:25:11.000             1.609688e+12  2021-01-03 15:25:06   

   purchaseDate_epoch         purchaseDate  pointsEarned  purchasedItemCount  \
0        1.609632e+12  2021-01-03 00:00:00         500.0                 5.0   
1        1.609601e+12  2021-01-02 15:24:43         150.0                 2.0   
2        1.609632e+12  2021-01-03 00:00:00           5.0                 1.0   
3        1.609632e+12  2021-01-03 00:00:00           5.0                 4.0   
4        1.609601e+12  2021-01-02 15:25:06           5.0                 2.0   

  rewardsReceiptStatus  totalSpent  
0             FINISHED        26.0  
1             FINISHED        11.0  
2             REJECTED        10.0  
3             FINISHED        28.0  
4             FINISHED         1.0

In [10]:
receipts_df.describe()

bonusPointsEarned  createDate_epoch  dateScanned_epoch  \
count         544.000000      1.119000e+03       1.119000e+03   
mean          238.893382      1.611800e+12       1.611800e+12   
std           299.091731      1.484091e+09       1.484091e+09   
min             5.000000      1.604089e+12       1.604089e+12   
25%             5.000000      1.610652e+12       1.610652e+12   
50%            45.000000      1.611941e+12       1.611941e+12   
75%           500.000000      1.612704e+12       1.612704e+12   
max           750.000000      1.614641e+12       1.614641e+12   

       finishedDate_epoch  ModifyDate_epoch  pointsAwardedDate_epoch  \
count        5.680000e+02      1.119000e+03             5.370000e+02   
mean         1.611058e+12      1.611847e+12             1.610949e+12   
std          9.534641e+08      1.361576e+09             1.056718e+09   
min          1.609687e+12      1.609687e+12             1.604089e+12   
25%          1.610141e+12      1.610660e+12             1.610043e+12   
50%          1.611091e+12      1.611941e+12             1.611090e+12   
75%          1.611769e+12      1.612704e+12             1.611710e+12   
max          1.614379e+12      1.614641e+12             1.614379e+12   

       purchaseDate_epoch  pointsEarned  purchasedItemCount   totalSpent  
count        6.710000e+02    609.000000           635.00000   684.000000  
mean         1.608548e+12    585.962890            14.75748    77.796857  
std          1.230755e+10   1357.166947            61.13424   347.110349  
min          1.509322e+12      0.000000             0.00000     0.000000  
25%          1.609805e+12      5.000000             1.00000     1.000000  
50%          1.610551e+12    150.000000             2.00000    18.200000  
75%          1.611624e+12    750.000000             5.00000    34.960000  
max          1.615225e+12  10199.800000           689.00000  4721.950000

In [4]:
receipts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   receipt_id               1119 non-null   object 
 1   user_id                  1119 non-null   object 
 2   bonusPointsEarned        544 non-null    float64
 3   bonusPointsEarnedReason  544 non-null    object 
 4   createDate_epoch         1119 non-null   int64  
 5   createDate               1119 non-null   object 
 6   dateScanned_epoch        1119 non-null   int64  
 7   dateScanned              1119 non-null   object 
 8   finishedDate_epoch       568 non-null    float64
 9   finishedDate             568 non-null    object 
 10  ModifyDate_epoch         1119 non-null   int64  
 11  modifyDate               1119 non-null   object 
 12  pointsAwardedDate_epoch  537 non-null    float64
 13  pointsAwardedDate        537 non-null    object 
 14  purchaseDate_epoch      

In [5]:
# Assumption: receipt_id should be unique in this table
receipts_df['receipt_id'].nunique() # this concludes that there are no duplicates in the table as well unless there same values are populated for different ids

1119

In [7]:
# Assumption: Total spent will not be a negative value since this table will only have order receipts and not the return receipts
receipts_df['totalSpent'].min()

0.0

Rewards funnel from a user journey perspective (left to right):

purchase date >> create date (purchasing leads to creating event) >> date scanned >> points awarded date >> finished date

Modify date can fall anywhere after the creat date in the funnle

In [12]:
# Convert date type for all dates
receipts_df['purchaseDate'] = pd.to_datetime(receipts_df['purchaseDate'], errors='coerce')
receipts_df['createDate'] = pd.to_datetime(receipts_df['createDate'], errors='coerce')
receipts_df['dateScanned'] = pd.to_datetime(receipts_df['dateScanned'], errors='coerce')
receipts_df['finishedDate'] = pd.to_datetime(receipts_df['finishedDate'], errors='coerce')
receipts_df['pointsAwardedDate'] = pd.to_datetime(receipts_df['pointsAwardedDate'], errors='coerce')
receipts_df['modifyDate'] = pd.to_datetime(receipts_df['modifyDate'], errors='coerce')

In [13]:
receipts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 20 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   receipt_id               1119 non-null   object        
 1   user_id                  1119 non-null   object        
 2   bonusPointsEarned        544 non-null    float64       
 3   bonusPointsEarnedReason  544 non-null    object        
 4   createDate_epoch         1119 non-null   int64         
 5   createDate               1119 non-null   datetime64[ns]
 6   dateScanned_epoch        1119 non-null   int64         
 7   dateScanned              1119 non-null   datetime64[ns]
 8   finishedDate_epoch       568 non-null    float64       
 9   finishedDate             568 non-null    datetime64[ns]
 10  ModifyDate_epoch         1119 non-null   int64         
 11  modifyDate               1119 non-null   datetime64[ns]
 12  pointsAwardedDate_epoch  537 non-n



1.   PurchaseDate - 671
2.   CreateDate - 1119
3.   dateScanned - 1119
4.   pointsAwardedDate - 537
5.   finishedDate - 568

**Every receipt_id will involve a purchaseDate, in this case we are missing records for almost half of the receipts data.**



In [18]:
validation_df = receipts_df.copy()
validation_df['pruchase_create'] = receipts_df['purchaseDate'] > receipts_df['createDate'] # returns a bool
validation_df['pruchase_create'].sum() # only counts when the value is True

13

There were 13 cases where purchase date is greater than create date which is an data inconsistency issue

In [19]:
validation_df['create_scanned'] = receipts_df['createDate'] > receipts_df['dateScanned']
validation_df['create_scanned'].sum()
# this is valid

0

In [20]:
validation_df['scanned_pointsawarded'] = receipts_df['dateScanned'] > receipts_df['pointsAwardedDate']
validation_df['scanned_pointsawarded'].sum()
# This is also valid

0

In [23]:
validation_df['pointsawarded_finished'] = receipts_df['pointsAwardedDate'] > receipts_df['finishedDate']
validation_df['pointsawarded_finished'].sum()
# This is also valid

0

In [21]:
validation_df['finished_create'] = receipts_df['createDate'] > receipts_df['pointsAwardedDate']
validation_df['finished_create'].sum()
# This is also valid

0

In [22]:
validation_df['finished_purchase'] = receipts_df['purchaseDate'] > receipts_df['pointsAwardedDate']
validation_df['finished_purchase'].sum()
# This is not valid it says points are awarded to users even before a user made their purchase

7

In [24]:
validation_df['finished_purchase'] = receipts_df['purchaseDate'] > receipts_df['finishedDate']
validation_df['finished_purchase'].sum()
# This is not valid it says process is finisged even before a user made their purchase

7

In [38]:
# Unusual Frequency of Records

validation_df['receipt_count'] = receipts_df.groupby('user_id')['receipt_id'].transform('count')
unusual_frequency = validation_df[validation_df['receipt_count'] > validation_df['receipt_count'].quantile(0.99)]
print("Unusual frequency of receipts per user:\n", unusual_frequency)

Unusual frequency of receipts per user:
 Empty DataFrame
Columns: [receipt_id, user_id, bonusPointsEarned, bonusPointsEarnedReason, createDate_epoch, createDate, dateScanned_epoch, dateScanned, finishedDate_epoch, finishedDate, ModifyDate_epoch, modifyDate, pointsAwardedDate_epoch, pointsAwardedDate, purchaseDate_epoch, purchaseDate, pointsEarned, purchasedItemCount, rewardsReceiptStatus, totalSpent, pruchase_create, create_scanned, scanned_pointsawarded, finished_create, finished_purchase, pointsawarded_finished, receipt_count_per_day, receipt_count]
Index: []

[0 rows x 28 columns]


While null values are common in the real world, there are some set busines processes/sequence the data must follow in order for Analyst to make inference from them.

Summary of the preliminary inconsistency check,
1. Every receipt should ideally have purchase date, total spent (assuming the table will only have order receipts and not the return receipts), purchasedItemCount, user_id
2. Based on the rewards funnel, purchase date should always be lesser than create date or records in later stage of the funnel, in this case there were 13 anomalies.

In [25]:
receipts_df.duplicated().value_counts()

False    1119
Name: count, dtype: int64

## Receipts_Item_List

In [26]:
uploaded = files.upload()
filename = list(uploaded.keys())[0]  # Geting the file name


receipts_item_df = pd.read_csv(filename)
receipts_item_df.head()

Saving receipt_list_df.csv to receipt_list_df.csv


user_id                receipt_id barcode     description  \
0  5ff1e1eacfcf6c399c274ae6  5ff1e1eb0a720f0523000575    4011  ITEM NOT FOUND   
1  5ff1e194b6a9d73a3a9f1052  5ff1e1bb0a720f052300056b    4011  ITEM NOT FOUND   
2  5ff1e1f1cfcf6c399c274b0b  5ff1e1f10a720f052300057a     NaN             NaN   
3  5ff1e1eacfcf6c399c274ae6  5ff1e1ee0a7214ada100056f    4011  ITEM NOT FOUND   
4  5ff1e194b6a9d73a3a9f1052  5ff1e1d20a7214ada1000561    4011  ITEM NOT FOUND   

   finalPrice  itemPrice needsFetchReview  partnerItemId  \
0        26.0       26.0            False            1.0   
1         1.0        1.0              NaN            1.0   
2         NaN        NaN            False            1.0   
3        28.0       28.0            False            1.0   
4         1.0        1.0              NaN            1.0   

  preventTargetGapPoints  quantityPurchased  ...  itemNumber  \
0                   True                5.0  ...         NaN   
1                    NaN                1.0  ...         NaN   
2                   True                NaN  ...         NaN   
3                   True                4.0  ...         NaN   
4                    NaN                1.0  ...         NaN   

  needsFetchReviewReason  originalMetaBriteQuantityPurchased  pointsEarned  \
0                    NaN                                 NaN           NaN   
1                    NaN                                 NaN           NaN   
2                    NaN                                 NaN           NaN   
3                    NaN                                 NaN           NaN   
4                    NaN                                 NaN           NaN   

   targetPrice competitiveProduct userFlaggedDescription deleted  \
0          NaN                NaN                    NaN     NaN   
1          NaN                NaN                    NaN     NaN   
2          NaN                NaN                    NaN     NaN   
3          NaN                NaN                    NaN     NaN   
4          NaN                NaN                    NaN     NaN   

  priceAfterCoupon metabriteCampaignId  
0              NaN                 NaN  
1              NaN                 NaN  
2              NaN                 NaN  
3              NaN                 NaN  
4              NaN                 NaN  

[5 rows x 34 columns]

In [28]:
receipts_item_df.describe()

finalPrice   itemPrice  partnerItemId  quantityPurchased  \
count  665.000000  665.000000     679.000000         665.000000   
mean    11.328421   11.328421     105.409426           1.488722   
std     10.985227   10.985227     309.048998           1.193712   
min      0.160000    0.160000       0.000000           1.000000   
25%      1.000000    1.000000       1.000000           1.000000   
50%      9.990000    9.990000       1.000000           1.000000   
75%     22.000000   22.000000       1.000000           1.000000   
max     50.000000   50.000000    1235.000000           8.000000   

       userFlaggedBarcode  userFlaggedPrice  userFlaggedQuantity  \
count        1.170000e+02        103.000000           103.000000   
mean         6.786334e+09         24.524272             3.135922   
std          2.229414e+10          3.127502             1.379461   
min          4.011000e+03         20.000000             1.000000   
25%          4.011000e+03         21.500000             2.000000   
50%          4.011000e+03         25.000000             3.000000   
75%          4.011000e+03         27.000000             4.000000   
max          7.940007e+10         29.000000             5.000000   

       originalMetaBriteBarcode  discountedItemPrice    itemNumber  \
count              1.400000e+01           199.000000  3.800000e+01   
mean               6.588450e+10            18.562513  9.806868e+09   
std                2.460178e+10            11.492729  2.045784e+10   
min                2.840064e+10             0.160000  4.023000e+03   
25%                4.151999e+10             4.975000  4.023000e+03   
50%                8.087804e+10            22.970000  4.023000e+03   
75%                8.087804e+10            28.570000  1.356230e+10   
max                8.087804e+10            50.000000  8.571830e+10   

       originalMetaBriteQuantityPurchased  pointsEarned  targetPrice  \
count                                 6.0    169.000000   119.000000   
mean                                  1.0     94.617751   763.546218   
std                                   0.0    166.094198   158.868816   
min                                   1.0      5.000000    77.000000   
25%                                   1.0      5.000000   800.000000   
50%                                   1.0     50.000000   800.000000   
75%                                   1.0    100.000000   800.000000   
max                                   1.0    870.000000   800.000000   

       userFlaggedDescription  priceAfterCoupon  
count                     0.0        100.000000  
mean                      NaN         24.776100  
std                       NaN          5.541729  
min                       NaN          0.790000  
25%                       NaN         22.970000  
50%                       NaN         25.770000  
75%                       NaN         28.570000  
max                       NaN         28.570000

In [29]:
receipts_item_df.info() # Shows the count and the null value count as well

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 34 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   user_id                             1119 non-null   object 
 1   receipt_id                          1119 non-null   object 
 2   barcode                             561 non-null    object 
 3   description                         584 non-null    object 
 4   finalPrice                          665 non-null    float64
 5   itemPrice                           665 non-null    float64
 6   needsFetchReview                    134 non-null    object 
 7   partnerItemId                       679 non-null    float64
 8   preventTargetGapPoints              117 non-null    object 
 9   quantityPurchased                   665 non-null    float64
 10  userFlaggedBarcode                  117 non-null    float64
 11  userFlaggedNewItem                  103 non

In [33]:
receipts_item_df.duplicated().value_counts() # No duplicates

False    1119
Name: count, dtype: int64

In [32]:
receipts_item_df['user_id'].nunique() # some users can have multiple purchases, user will have 1-to-Many relation

258

In [31]:
receipts_item_df['barcode'].nunique() # value is far lower than expected

252

In [34]:
receipts_item_df['brandCode'].nunique() # value is far lower than expected

18

In [35]:
receipts_item_df['partnerItemId'].nunique() # value is far lower than expected

28

In [36]:
receipts_item_df['rewardsProductPartnerId'].nunique() # value is far lower than expected

8

Summary:

1. Based on the receipts table, we can infer that receipt_id will be unique.

2. Every product has a unique barcode associated with them, this table only has records for 50% of the table which is a **Data Inconsistency**.
3. Every item should also have other important records such as itemId, brandCode, user_id, product partner id, and so on, which does not seem to be the case here so a huge **Data Inconsistency**.

Overall this table has lot of null values, on average a volume of less than 10-15% in missing value is acceptable but here that volume is far greater. Hence, the data collection process for this table needs reconfiguration before infering anything from this table.

## Users

In [41]:
uploaded = files.upload()
filename = list(uploaded.keys())[0]  # Geting the file name


users_df = pd.read_csv(filename)
users_df.head()

Saving users.csv to users.csv


active  createdDate_epoch  lastLogin_epoch      role signUpSource state  \
0    True      1609687444800     1.609688e+12  consumer        Email    WI   
1    True      1609687444800     1.609688e+12  consumer        Email    WI   
2    True      1609687444800     1.609688e+12  consumer        Email    WI   
3    True      1609687530554     1.609688e+12  consumer        Email    WI   
4    True      1609687444800     1.609688e+12  consumer        Email    WI   

                    user_id              createdDate                lastLogin  
0  5ff1e194b6a9d73a3a9f1052  2021-01-03 15:24:04.800  2021-01-03 15:25:37.858  
1  5ff1e194b6a9d73a3a9f1052  2021-01-03 15:24:04.800  2021-01-03 15:25:37.858  
2  5ff1e194b6a9d73a3a9f1052  2021-01-03 15:24:04.800  2021-01-03 15:25:37.858  
3  5ff1e1eacfcf6c399c274ae6  2021-01-03 15:25:30.554  2021-01-03 15:25:30.597  
4  5ff1e194b6a9d73a3a9f1052  2021-01-03 15:24:04.800  2021-01-03 15:25:37.858

In [42]:
users_df.describe()

createdDate_epoch  lastLogin_epoch
count       4.950000e+02     4.330000e+02
mean        1.596678e+12     1.611388e+12
std         4.402970e+10     4.430109e+09
min         1.418999e+12     1.525714e+12
25%         1.609789e+12     1.610130e+12
50%         1.610569e+12     1.611237e+12
75%         1.611596e+12     1.612366e+12
max         1.613139e+12     1.614963e+12

In [43]:
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   active             495 non-null    bool   
 1   createdDate_epoch  495 non-null    int64  
 2   lastLogin_epoch    433 non-null    float64
 3   role               495 non-null    object 
 4   signUpSource       447 non-null    object 
 5   state              439 non-null    object 
 6   user_id            495 non-null    object 
 7   createdDate        495 non-null    object 
 8   lastLogin          433 non-null    object 
dtypes: bool(1), float64(1), int64(1), object(6)
memory usage: 31.5+ KB


In [44]:
users_df.duplicated().value_counts() # More than half the records (283) in the user table is a duplicate

True     283
False    212
Name: count, dtype: int64

In [47]:
# Funnel from a user account perspective: Created date >> last login date
# Convert date type for all dates
users_df['createdDate'] = pd.to_datetime(users_df['createdDate'], errors='coerce')
users_df['lastLogin'] = pd.to_datetime(users_df['lastLogin'], errors='coerce')

validation_users_df = users_df.copy()
validation_users_df['create_login'] = users_df['createdDate'] > users_df['lastLogin']
validation_users_df['create_login'].sum()

# this is a valid sequence

0

In [48]:
users_drop = users_df.drop_duplicates()
users_drop.info()

<class 'pandas.core.frame.DataFrame'>
Index: 212 entries, 0 to 475
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   active             212 non-null    bool          
 1   createdDate_epoch  212 non-null    int64         
 2   lastLogin_epoch    172 non-null    float64       
 3   role               212 non-null    object        
 4   signUpSource       207 non-null    object        
 5   state              206 non-null    object        
 6   user_id            212 non-null    object        
 7   createdDate        212 non-null    datetime64[ns]
 8   lastLogin          172 non-null    datetime64[ns]
dtypes: bool(1), datetime64[ns](2), float64(1), int64(1), object(4)
memory usage: 15.1+ KB


Summary:

This table has lots of duplicates, after droping them the count of rows came down to 212 unique records. For this deduped table, the records seem to be valid with acceptable volume of null values (optionally, can work with the DE team to remove those nulls as well)

One business caveat to ensure that only users active during the time of purchase can make purchases.

## Brands

In [49]:
uploaded = files.upload()
filename = list(uploaded.keys())[0]  # Geting the file name


brand_df = pd.read_csv(filename)
brand_df.head()

Saving brands.csv to brands.csv


barcode        category      categoryCode                       name  \
0  511111019862          Baking            BAKING  test brand @1612366101024   
1  511111519928       Beverages         BEVERAGES                  Starbucks   
2  511111819905          Baking            BAKING  test brand @1612366146176   
3  511111519874          Baking            BAKING  test brand @1612366146051   
4  511111319917  Candy & Sweets  CANDY_AND_SWEETS  test brand @1612366146827   

   topBrand                      brandCode                  brand_id  \
0       0.0                            NaN  601ac115be37ce2ead437551   
1       0.0                      STARBUCKS  601c5460be37ce2ead43755f   
2       0.0  TEST BRANDCODE @1612366146176  601ac142be37ce2ead43755d   
3       0.0  TEST BRANDCODE @1612366146051  601ac142be37ce2ead43755a   
4       0.0  TEST BRANDCODE @1612366146827  601ac142be37ce2ead43755e   

                     cpg_id cpg_ref  
0  601ac114be37ce2ead437550    Cogs  
1  5332f5fbe4b03c9a25efd0ba    Cogs  
2  601ac142be37ce2ead437559    Cogs  
3  601ac142be37ce2ead437559    Cogs  
4  5332fa12e4b03c9a25efd1e7    Cogs

In [50]:
brand_df.describe()

barcode    topBrand
count  1.167000e+03  555.000000
mean   5.111115e+11    0.055856
std    2.874497e+05    0.229850
min    5.111110e+11    0.000000
25%    5.111112e+11    0.000000
50%    5.111114e+11    0.000000
75%    5.111117e+11    0.000000
max    5.111119e+11    1.000000

In [51]:
brand_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   barcode       1167 non-null   int64  
 1   category      1012 non-null   object 
 2   categoryCode  517 non-null    object 
 3   name          1167 non-null   object 
 4   topBrand      555 non-null    float64
 5   brandCode     898 non-null    object 
 6   brand_id      1167 non-null   object 
 7   cpg_id        1167 non-null   object 
 8   cpg_ref       1167 non-null   object 
dtypes: float64(1), int64(1), object(7)
memory usage: 82.2+ KB


In [52]:
brand_df.duplicated().value_counts() # there are no duplicate records

False    1167
Name: count, dtype: int64

In [53]:
brand_df['barcode'].nunique() # there are 7 records with duplicate barcode, barcodes are unique for every product

1160

In [54]:
brand_df['brandCode'].nunique() # brandCode should exist for every brand, this is a data inconsistency

896

In [55]:
brand_df['cpg_id'].nunique() # connect with SME to learn more about this id and check for its uniqueness

196

Summary:

1. This table does not have any duplicates, that is a good use of storage resources.
2. Barcode should be unique for every item and a brand can have multiple items/barcodes associated with them. This is a key to connect with receipts_item_list table. The table only has barcode for 77% of the records which is a **data inconsistency** considering this column is ideally a non-null column.
3. Ideally brandCode count should be same brand name count or brand_id count but that does not seem to be the case which is a **data inconsistency**.
4. Need to clarify on the cpg_id column from the subject matter expert.

# Overall Conclusion

This is a preliminary inconsistency report focused on validity of the data itself.

Based on the problem statement, the range of data consistency check differs, for example, if this data is used to build a machine learning model then the validity check will include outlier detection, correlation check, data distribution check, and so on.